File: backend/app/services/integrations.py
from __future__ import annotations

from fastapi import Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.logging import get_logger
from app.db.models.enums import IntegrationStatus
from app.db.models.integration import Integration
from app.dependencies import db_session
from app.models.integrations import (
    ConnectionTestResult,
    IntegrationCreate,
    IntegrationRead,
    IntegrationUpdate,
)
from app.providers.anthropic_client import anthropic_client
from app.providers.elevenlabs import elevenlabs_client
from app.providers.lipsync import lipsync_client
from app.providers.push import push_client
from app.providers.storage import storage
from app.providers.twilio_client import twilio_client
from app.services.base import Service

log = get_logger(__name__)


async def _probe(provider: str) -> tuple[bool, str]:
    p = provider.lower()
    try:
        if p == "twilio":
            if not twilio_client.configured:
                return False, "Missing account SID / auth token / phone number."
            return True, f"Ready · {settings.twilio_phone_number}"
        if p == "elevenlabs":
            if not elevenlabs_client.configured:
                return False, "ELEVENLABS_API_KEY is not set."
            acct = await elevenlabs_client.account()
            tier = acct.get("tier", "active")
            agent = " · agent set" if elevenlabs_client.agent_configured else " · no agent id"
            return True, f"Connected · {tier}{agent}"
        if p in {"anthropic", "claude"}:
            return (
                (True, f"Key present · {settings.anthropic_chat_model}")
                if anthropic_client.configured
                else (False, "ANTHROPIC_API_KEY is not set.")
            )
        if p in {"heygen", "did", "lipsync"}:
            return (
                (True, f"Ready · {lipsync_client.provider}")
                if lipsync_client.configured
                else (False, f"{lipsync_client.provider} not fully configured.")
            )
        if p in {"s3", "aws", "storage"}:
            if not storage.configured:
                return False, "No bucket configured."
            ok = await storage.health()
            return (ok, "Bucket reachable." if ok else "Bucket not reachable.")
        if p in {"push", "webpush"}:
            return (
                (True, "VAPID keys present.")
                if push_client.configured
                else (False, "VAPID keys are not configured.")
            )
    except Exception as exc:
        return False, f"{exc}"
    return False, "No connection test for this provider."


class IntegrationService(Service):
    async def list_integrations(self) -> list[IntegrationRead]:
        rows = (
            (await self.session.execute(select(Integration).order_by(Integration.name)))
            .scalars()
            .all()
        )
        return [IntegrationRead.model_validate(r) for r in rows]

    async def get(self, integration_id: str) -> IntegrationRead:
        return IntegrationRead.model_validate(
            await self._get(Integration, integration_id, label="Integration")
        )

    async def create(self, data: IntegrationCreate) -> IntegrationRead:
        row = Integration(
            name=data.name,
            provider=data.provider,
            type=data.type,
            status=data.status,
            configuration=data.configuration,
        )
        self.session.add(row)
        await self.session.flush()
        return IntegrationRead.model_validate(row)

    async def update(self, integration_id: str, data: IntegrationUpdate) -> IntegrationRead:
        row = await self._get(Integration, integration_id, label="Integration")
        for field, value in data.model_dump(exclude_unset=True).items():
            setattr(row, field, value)
        await self.session.flush()
        return IntegrationRead.model_validate(row)

    async def test_connection(self, integration_id: str) -> ConnectionTestResult:
        row = await self._get(Integration, integration_id, label="Integration")
        ok, detail = await _probe(row.provider)
        row.status = IntegrationStatus.CONNECTED if ok else IntegrationStatus.ERROR
        await self.session.flush()
        return ConnectionTestResult(ok=ok, status=row.status, detail=detail)


def get_integrations_service(
    session: AsyncSession = Depends(db_session),
) -> IntegrationService:
    return IntegrationService(session)

File: apps/web/src/lib/integrations.ts
/** Integrations API calls. Types mirror backend/app/models/integrations.py.
 *
 * `configuration` (which may hold secrets) is intentionally never returned by
 * the API — `IntegrationRead` doesn't have the field at all. The credential
 * UI is built around that: it can only ever *set/replace* a field, never
 * show or diff against a real stored value.
 */
import { apiFetch } from "./api-client";

export type IntegrationStatus = "connected" | "error" | "disabled";

export interface IntegrationRead {
  id: string;
  name: string;
  provider: string;
  type: string | null;
  status: IntegrationStatus;
  created_at: string;
}

export interface IntegrationCreate {
  name: string;
  provider: string;
  type?: string;
  configuration?: Record<string, unknown>;
  status?: IntegrationStatus;
}

export interface IntegrationUpdate {
  name?: string;
  type?: string;
  status?: IntegrationStatus;
  configuration?: Record<string, unknown>;
}

export interface ConnectionTestResult {
  ok: boolean;
  status: IntegrationStatus;
  detail: string | null;
}

export function listIntegrations(): Promise<IntegrationRead[]> {
  return apiFetch<IntegrationRead[]>("/api/integrations");
}

export function getIntegration(id: string): Promise<IntegrationRead> {
  return apiFetch<IntegrationRead>(`/api/integrations/${id}`);
}

/** Admin only. Not currently exposed by any UI flow — there is no "add
 * integration" screen, only configuring the fixed set of rows the backend
 * seeds — but kept here so one can be added without touching the backend
 * contract. */
export function createIntegration(
  input: IntegrationCreate,
): Promise<IntegrationRead> {
  return apiFetch<IntegrationRead>("/api/integrations", {
    method: "POST",
    json: input,
  });
}

/** Admin only (backend returns 403 for non-admins). */
export function updateIntegration(
  id: string,
  patch: IntegrationUpdate,
): Promise<IntegrationRead> {
  return apiFetch<IntegrationRead>(`/api/integrations/${id}`, {
    method: "PATCH",
    json: patch,
  });
}

/** Admin only. Live-probes the real provider and updates + returns the
 * row's status based on the result — not a simulation. */
export function testConnection(id: string): Promise<ConnectionTestResult> {
  return apiFetch<ConnectionTestResult>(`/api/integrations/${id}/test`, {
    method: "POST",
  });
}

/* ------------------------------------------------------------------------ *
 * Provider catalog — client-side UI metadata only.
 *
 * The backend's `configuration` column is an opaque JSON object; it has no
 * concept of "fields." This is a static description of the shape each known
 * provider's configuration takes (matching the provider strings the
 * backend's connection-test probe matches on — see
 * backend/app/services/integrations.py:_probe), used only to render a
 * sensible credential form. It's a UI convenience describing a known
 * integration, not fabricated data: no values, statuses, or "configured"
 * flags here, only field labels/keys.
 * ------------------------------------------------------------------------ */

export interface ProviderField {
  key: string;
  label: string;
  secret: boolean;
  hint?: string;
}

export interface ProviderCapability {
  id: string;
  label: string;
}

export interface ProviderMeta {
  category: string;
  purpose: string;
  description: string;
  docsUrl: string;
  icon: "phone" | "waves" | "sparkles" | "clapperboard" | "database" | "bell";
  capabilities: ProviderCapability[];
  fields: ProviderField[];
}

const PROVIDER_CATALOG: Record<string, ProviderMeta> = {
  twilio: {
    category: "Telephony",
    purpose: "Phone communication",
    description:
      "Programmable Voice for inbound and outbound calls, the phone number, and call state.",
    docsUrl: "https://www.twilio.com/docs/voice",
    icon: "phone",
    capabilities: [
      { id: "inbound", label: "Inbound calls" },
      { id: "outbound", label: "Outbound calls" },
      { id: "webhook-validation", label: "Webhook signature validation" },
    ],
    fields: [
      { key: "account_sid", label: "Account SID", secret: true },
      { key: "auth_token", label: "Auth token", secret: true },
      { key: "phone_number", label: "Phone number", secret: false },
    ],
  },
  elevenlabs: {
    category: "Voice",
    purpose: "AI voice and voice clone",
    description:
      "Text-to-speech, speech-to-text, and the Professional Voice Clone used on every call.",
    docsUrl: "https://elevenlabs.io/docs",
    icon: "waves",
    capabilities: [
      { id: "tts", label: "Text-to-speech" },
      { id: "stt", label: "Speech-to-text" },
      { id: "clone", label: "Professional Voice Clone" },
      { id: "agent", label: "Voice agent" },
    ],
    fields: [
      { key: "api_key", label: "API key", secret: true },
      { key: "agent_id", label: "Agent ID", secret: false },
      {
        key: "voice_id",
        label: "Voice ID (Professional Voice Clone)",
        secret: false,
        hint: "Must be the configured Professional Voice Clone — not a stock voice.",
      },
    ],
  },
  anthropic: {
    category: "AI engine",
    purpose: "AI reasoning engine",
    description:
      "Claude powers the dashboard chat, the daily-report research + script, and grading.",
    docsUrl: "https://docs.anthropic.com",
    icon: "sparkles",
    capabilities: [
      { id: "chat", label: "Chat completions" },
      { id: "streaming", label: "SSE streaming" },
      { id: "web-search", label: "Web search (market reports)" },
    ],
    fields: [
      { key: "api_key", label: "API key", secret: true },
      { key: "chat_model", label: "Chat model", secret: false },
    ],
  },
  heygen: {
    category: "Video",
    purpose: "Avatar video generation",
    description:
      "Talking-head avatar rendering for the daily market report. Runs in the worker, never in a request.",
    docsUrl: "https://docs.heygen.com",
    icon: "clapperboard",
    capabilities: [
      { id: "avatar-video", label: "Avatar video" },
      { id: "talking-head", label: "Talking-head render" },
    ],
    fields: [
      { key: "api_key", label: "API key", secret: true },
      { key: "avatar_id", label: "Avatar ID", secret: false, hint: "The avatar to render the report with." },
    ],
  },
  s3: {
    category: "Storage",
    purpose: "Media storage",
    description:
      "Object storage for call recordings, generated audio, avatar videos, captions, and training exports.",
    docsUrl: "https://docs.aws.amazon.com/s3",
    icon: "database",
    capabilities: [
      { id: "recordings", label: "Call recordings" },
      { id: "audio", label: "Generated audio" },
      { id: "video", label: "Avatar videos" },
      { id: "exports", label: "Training exports" },
    ],
    fields: [
      { key: "access_key_id", label: "Access key ID", secret: true },
      { key: "secret_access_key", label: "Secret access key", secret: true },
      { key: "region", label: "Region", secret: false, hint: "e.g. us-east-1" },
      { key: "bucket", label: "Bucket", secret: false },
    ],
  },
  push: {
    category: "Notifications",
    purpose: "Browser push notifications",
    description: "Web Push (VAPID) notifications for operator alerts.",
    docsUrl: "https://developer.mozilla.org/en-US/docs/Web/API/Push_API",
    icon: "bell",
    capabilities: [{ id: "webpush", label: "Web push delivery" }],
    fields: [
      { key: "vapid_public_key", label: "VAPID public key", secret: false },
      { key: "vapid_private_key", label: "VAPID private key", secret: true },
      { key: "vapid_subject", label: "Subject (mailto:)", secret: false },
    ],
  },
};

/** Aliases the backend probe also matches on (see `_probe` in
 * backend/app/services/integrations.py), mapped to the canonical catalog
 * entry above. */
const PROVIDER_ALIASES: Record<string, string> = {
  claude: "anthropic",
  did: "heygen",
  lipsync: "heygen",
  aws: "s3",
  storage: "s3",
  webpush: "push",
};

const GENERIC_META: ProviderMeta = {
  category: "Integration",
  purpose: "Third-party service",
  description: "Configuration for this integration is a freeform key/value object.",
  docsUrl: "#",
  icon: "sparkles",
  capabilities: [],
  fields: [],
};

/** Looks up the static field/capability catalog for a provider string,
 * matched case-insensitively the same way the backend's connection test
 * does. Returns a generic (fields-less) fallback for unknown providers —
 * those get a generic key/value editor in the UI instead of named fields. */
export function getProviderMeta(provider: string): ProviderMeta {
  const key = provider.trim().toLowerCase();
  const canonical = PROVIDER_ALIASES[key] ?? key;
  return PROVIDER_CATALOG[canonical] ?? GENERIC_META;
}

export function isKnownProvider(provider: string): boolean {
  const key = provider.trim().toLowerCase();
  const canonical = PROVIDER_ALIASES[key] ?? key;
  return canonical in PROVIDER_CATALOG;
}
